fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/packet/DirectRegistryPacketHandler.java

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.registry.sync.packet;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import com.google.common.base.Preconditions;
+import io.netty.buffer.Unpooled;
+import it.unimi.dsi.fastutil.objects.Object2IntLinkedOpenHashMap;
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.network.PacketByteBuf;
+import net.minecraft.server.network.ServerPlayerEntity;
+import net.minecraft.util.Identifier;
+
+/**
+ * A more optimized method to sync registry ids to client.
+ * Produce smaller packet than old {@link NbtRegistryPacketHandler nbt-based} method.
+ *
+ * <p>This method optimize the packet in multiple way:
+ * <ul>
+ *     <li>Directly write into the buffer instead of using an nbt;</li>
+ *     <li>Group all {@link Identifier} with same namespace together and only send those unique namespaces once for each group;</li>
+ *     <li>Group adjacent rawIds together and only send the difference of the first rawId and the last rawId of the bulk before.
+ *     This is based on the assumption that mods generally register all of their object at once,
+ *     therefore making the rawIds somewhat densely packed.</li>
+ * </ul>
+ */
+public class DirectRegistryPacketHandler extends RegistryPacketHandler {
+	public static final RegistryPacketHandler INSTANCE = new DirectRegistryPacketHandler();
```

Comment:
Small nitpick: With `DirectRegistryPacketHandler` now being stateful, it should not be singleton-ish. The `DirectRegistryPacketHandler` instance used at runtime should probably be placed elsewhere, such as in a static field in `RegistrySyncManager`.

Reply:
Done :+1: